demos/remote_attestation/maa/gen_quote/gen_maa_json.c

```diff
@@ -0,0 +1,115 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <openssl/sha.h>
+
+#include "sgx_quote_3.h"
+#include "occlum_dcap.h"
+
+#define MAA_JSON "/host/maa.json"
+
+void sha256sum(const uint8_t *data, uint32_t data_size, uint8_t *hash)
+{
+    SHA256_CTX sha256;
+    SHA256_Init(&sha256);
+    SHA256_Update(&sha256, data, data_size);
+    SHA256_Final(hash, &sha256);
+}
+
+const char *uint16_to_buffer (char *buffer, unsigned int maxSize, uint16_t n, size_t size)
+{
+    if (size * 2 >= maxSize || size < 2)
+        return "DEADBEEF";
+    sprintf(&buffer[0], "%02X", (uint8_t)(n));
+    sprintf(&buffer[2], "%02X", (uint8_t)(n >> 8));
+
+    for (int i=2; i < size; i++)
+    {
+        sprintf(&buffer[i*2], "%02X", 0);
+    }
+    buffer[size*2+1] = '\0';
+    return buffer;
+}
+
+const char *format_hex_buffer (char *buffer, unsigned int maxSize, uint8_t *data, size_t size)
+{
+    if (size * 2 >= maxSize)
+        return "DEADBEEF";
+
+    for (int i=0; i < size; i++)
+    {
+        sprintf(&buffer[i*2], "%02X", data[i]);
+    }
+    buffer[size*2+1] = '\0';
+    return buffer;
+}
+
+
+void main() {
+    void *handle;
+    uint32_t quote_size;
+    uint8_t *p_quote_buffer;
+    sgx_quote3_t *p_quote;
+    sgx_report_body_t *p_report_body;
+    int32_t ret;
+
+    const int hex_buffer_size = 1024*64;
+    char hex_buffer[hex_buffer_size];
+    
+    handle = dcap_quote_open();
+    quote_size = dcap_get_quote_size(handle);
+    printf("quote size = %d\n", quote_size);
+
+    p_quote_buffer = (uint8_t*)malloc(quote_size);
+    if (NULL == p_quote_buffer) {
+        printf("Couldn't allocate quote_buffer\n");
+        goto CLEANUP;
+    }
+    memset(p_quote_buffer, 0, quote_size);
+
+    uint8_t enclave_held_data[6] = {0x01, 0x02, 0x03, 0x04, 0x05, 0x06};
```

Comment:
Is there an example somewhere of how this can be retrieved/computed from an app running with Occlum, for example a JVM-based app?

Reply:
@tychobrailleur Did you mean how to retrive the "enclave_held_data"? 
The "enclave_held_data" in this example is actaully sgx report data which usually is input by the user.
Once the sgx quote is generated, the report data could be parsed from it, just like the example does in https://github.com/qzheng527/occlum/blob/maa/demos/remote_attestation/maa/gen_quote/gen_maa_json.c#L105.

So for a JVM-based app running in Occlum, similar DCAP operation needs to be added to retrieve any information from the SGX quote. And if the DCAP operation is not feasible to add in the app, Occlum recomends to add it in the Occlum init stage. 
```
For every Occlum built application, it starts [init](https://github.com/occlum/occlum/blob/master/tools/init) process first, then starts the real application in RootFS. The default init process just run RootFS integrity check and then load the RootFS where the real application is located.
```
Thus a customized init needs to be provided when building Occlum image. For how to customize the init , please refer to the demo [`init_ra_flow`](https://github.com/occlum/occlum/tree/master/demos/remote_attestation/init_ra_flow).

Reply:
Yes, that's what I meant.  Makes sense, that `occlum init` info is helpful!  thanks @qzheng527